src/microbit/model/display.py

```diff
@@ -58,13 +63,20 @@ def scroll(self, value, delay=150, wait=True, loop=False, monospace=False):
         while True:
             # Show the scrolled image one square at a time.
             for x in range(appended_image.width() - CONSTANTS.LED_WIDTH + 1):
+                self.__image_lock.acquire()
                 self.__image.blit(
                     appended_image, x, 0, CONSTANTS.LED_WIDTH, CONSTANTS.LED_HEIGHT
                 )
+                self.__image_lock.release()
+
                 # If show or scroll is called again, there will be a different pid and break
+                self.__thread_identifier_lock.acquire()
                 if self.__current_pid != threading.get_ident():
                     break
+                self.__thread_identifier_lock.release()
+
                 time.sleep(delay / 1000)
+

```

Comment:
This method can still go wrong.
For e.g.: If we called `scroll(wait=False)` and after some time `show(wait=True)`, the following could happen:
`show` changes the image, `scroll` changes the image, `show` changes the pid, `scroll` changes the pid. On the screen it's a snapshot of scroll, instead of the image from show.
You could solve this by having one lock and doing the both pid and image checks/changes inside of it.

Reply:
When you say 

> You could solve this by having one lock and doing the both pid and image checks/changes inside of it.

Do you mean have everything change in one block between acquire and release?

Reply:
Looks good to me in the last change.